hoot-core-test/src/test/cpp/hoot/core/util/HttpTestServer.cpp

```diff
@@ -181,22 +188,49 @@ bool HttpTestServer::respond(HttpConnection::HttpConnectionPtr& connection)
   return !_interupt;
 }
 
-std::string HttpTestServer::read_request_headers(HttpConnection::HttpConnectionPtr& connection)
+void HttpTestServer::parse_request(HttpConnection::HttpConnectionPtr &connection)
 {
+  //  Reset headers and body
+  _headers.clear();
+  _body.clear();
   //  Read the HTTP request headers
-  boost::asio::streambuf buf;
-  boost::asio::read_until(connection->socket(), buf, "\r\n\r\n");
-  return boost::asio::buffer_cast<const char*>(buf.data());
-}
-
-std::string HttpTestServer::read_request_body(const std::string& headers, HttpConnection::HttpConnectionPtr& connection)
-{
-  //  Parse the headers to get the content length
+  boost::asio::streambuf stream_buf;
+  size_t bytes_read = boost::asio::read_until(connection->socket(), stream_buf, "\r\n\r\n");
+  std::string headers = boost::asio::buffer_cast<const char*>(stream_buf.data());
+  std::string::size_type pos = headers.find("\r\n\r\n");
+  if (pos == std::string::npos)
+  {
+    //  Throw some error here
+  }
+  //  Parse the content length
   long content_length = parse_content_length(headers);
+  if (content_length == 0)
+  {
+    _headers = headers;
+    return;
+  }
+  //  Create a buffer for the entire contents
   std::vector<char> buf(content_length + 1, 0);
-  //  Read the HTTP request body
-  boost::asio::read(connection->socket(), boost::asio::buffer(buf, content_length));
-  return buf.data();
+  //  When only the headers are read, read all of the contents
+  if (headers.length() == bytes_read)
+  {
+    //  Set the headers
+    _headers = headers;
+    //  Read the HTTP request body
+    boost::asio::read(connection->socket(), boost::asio::buffer(buf, content_length));
+    _body = buf.data();
+  }
+  else
+  {
+    //  Headers (for some reason) contains part of the body already read
+    //  copy that section to the body
+    _body = headers.substr(pos + 4);
+    long read_length = content_length - (headers.length() - pos - 4);
+    boost::asio::read(connection->socket(), boost::asio::buffer(buf, read_length));
+    _body.append(buf.data());
+    //  Finally update the headers
+    _headers = headers.substr(0, pos);
+  }
```

Comment:
Sometimes more is sent to the buffer in a `read_until` even though the name says it will only "read until" the first `\r\n\r\n` break is found.  So the read code has to deal with a read overrun to get the body and parse it correctly.  Before this was causing occasional hang-ups in unit tests (and potentially production) that were timing out.